_posts/2024-11-13-boosting-k-nn-exact-search.md

```diff
@@ -0,0 +1,123 @@
+
+```json
+PUT my-knn-index-1
+{
+  "mappings": {
+    "properties": {
+      "my_vector1": { "type": "knn_vector", "dimension": 2 },
+      "my_vector2": { "type": "knn_vector", "dimension": 4 }
+    }
+  }
+}
+```
+
+Next, index some sample data:
+
+```json
+POST _bulk
+{ "index": { "_index": "my-knn-index-1", "_id": "1" } }
+{ "my_vector1": [1.5, 2.5], "price": 12.2 }
+{ "index": { "_index": "my-knn-index-1", "_id": "2" } }
+{ "my_vector1": [2.5, 3.5], "price": 7.1 }
+// Additional documents omitted for brevity
+```
+
+Finally, run an exact k-NN search using the `script_score` query:
+
+```json
+GET my-knn-index-1/_search
+{
+ "size": 4,
+ "query": {
+   "script_score": {
+     "query": { "match_all": {} },
+     "script": {
+       "source": "knn_score",
+       "lang": "knn",
+       "params": {
+         "field": "my_vector2",
+         "query_value": [2.0, 3.0, 5.0, 6.0],
+         "space_type": "l2"
+       }
+     }
+   }
+ }
+}
+```
+
+For more information about exact k-NN search and `script_score` queries, see [Exact k-NN with scoring script](https://opensearch.org/docs/latest/search-plugins/knn/knn-score-script/). You'll find detailed guides to help you configure k-NN exact search and make the most of custom scoring.
+
+## Experiments show real-world performance gains
+
+To measure the impact of these optimizations, we conducted A/B tests comparing OpenSearch 2.14 to OpenSearch 2.15 and 2.17 in a single-node cluster. 
+
+### Cluster configuration
+
+|Dataset	|Cohere 1m	|
+|---	|---	|
+|Data nodes	|1	|
+|CPUs	|8	|
+|EBS Volume (GB)	|500	|
+
+### Results
+
+The following table provides latency comparison between OpenSearch versions 2.14 and 2.15.
+
+|Space type	|Version	|50th percentile latency (ms)	|90th percentile latency (ms)	|99th percentile latency (ms)	|
+|---	|---	|---	|---	|---	|
+|**Inner product**	|2.14	|668.84378	|816.95014	|948.21019	|
+|| 2.15	|481.71112	|499.61605	|519.03619	|
+|| **Improvement**	|27.98%	|38.84%	|45.26%	|
+|**L2**	|2.14	|670.98628	|682.84925	|693.12135	|
+|| 2.15	|503.96816	|520.86304	|537.51321	|
+|| **Improvement**	|24.89%	|23.72%	|22.45%	|
+
+The following table provides latency comparison between OpenSearch versions 2.14 and 2.17.
+
+|Space type	|Version	|50th percentile latency (ms)	|90th percentile latency (ms)	|99th percentile latency (ms)	|
+|---	|---	|---	|---	|---	|
+|**Inner product**	|2.14	|668.84378	|816.95014	|948.21019	|
+||2.17	|99.58072	|117.0792	|121.36626	|
+||**Improvement**	|85.11%	|85.67%	|87.20%	|
+|**L2**|2.14	|670.98628	|682.84925	|693.12135	|
+| | 2.17	|104.36596	|118.85475	|127.60656	|
+|| **Improvement**	|84.45%	|82.59%	|81.59%	|
+
+### Conclusion
+
+The tests showed that OpenSearch's new SIMD optimizations resulted in significant latency reductions, especially for the inner product space type, which saw up to an 87% latency reduction at the 99th percentile.
+
+## What's next for exact k-NN search?
+
```

Comment:
Line 123: In the second sentence, rather than "directly" preceding "at query time", should it instead precede "switch"?

Reply:
I think we can remove "directly". Let me do that.